src/command_modules/azure-cli-hdinsight/azure/cli/command_modules/hdinsight/custom.py

```diff
@@ -0,0 +1,238 @@
+        raise CLIError('If storage details are specified, the storage account, storage account key, and either the default container or default filesystem must be specified.')
+
+    # Validate network profile parameters
+    if not _all_or_none(virtual_network, subnet_name):
+        raise CLIError('Either both the virtual network and subnet should be specified, or neither should be.')
+    # Specify virtual network profile only when network arguments are provided
+    virtual_network_profile = virtual_network and VirtualNetworkProfile(
+        id=virtual_network,
+        subnet=subnet_name
+    )
+
+    # Validate data disk parameters
+    if not workernode_data_disks_per_node and workernode_data_disk_storage_account_type:
+        raise CLIError("Cannot define data disk storage account type unless disks per node is defined.")
+    if not workernode_data_disks_per_node and workernode_data_disk_size:
+        raise CLIError("Cannot define data disk size unless disks per node is defined.")
+    # Specify data disk groups only when disk arguments are provided
+    workernode_data_disk_groups = workernode_data_disks_per_node and [
+        DataDisksGroups(
+            disks_per_node=workernode_data_disks_per_node,
+            storage_account_type=workernode_data_disk_storage_account_type,
+            disk_size_gb=workernode_data_disk_size
+        )
+    ]
+
+    os_profile = OsProfile(
+        linux_operating_system_profile=LinuxOperatingSystemProfile(
+            username=ssh_username,
+            password=ssh_password,
+            ssh_public_key=ssh_public_key
+        )
+    )
+
+    create_params = ClusterCreateParametersExtended(
+        location=location,
+        tags=tags,
+        properties=ClusterCreateProperties(
+            cluster_version=cluster_version,
+            os_type=OSType.linux,
+            tier=cluster_tier,
+            cluster_definition=ClusterDefinition(
+                kind=cluster_type,
+                configurations=cluster_configurations,
+                component_version=component_version
+            ),
+            compute_profile=ComputeProfile(
+                roles=[
+                    # Required roles
+                    Role(
+                        name="headnode",
+                        target_instance_count=2,
+                        hardware_profile=HardwareProfile(vm_size=headnode_size),
+                        os_profile=os_profile,
+                        virtual_network_profile=virtual_network_profile
+                    ),
+                    Role(
+                        name="workernode",
+                        target_instance_count=workernode_count,
+                        hardware_profile=HardwareProfile(vm_size=workernode_size),
+                        os_profile=os_profile,
+                        virtual_network_profile=virtual_network_profile,
+                        data_disks_groups=workernode_data_disk_groups
+                    )
+                ] +
```

Comment:
Why do you use `+` operator here instead of just initialize the list altogether?

Reply:
We want the list to contain certain elements under certain conditions. Here, only want to define zookeepernode/edgenode roles if it was specified by the user. The alternative approach would be to initialize a list beforehand, and add to it if necessary- we can take that approach instead if that's preferred.

Reply:
This statement is too long. Cut it into smaller pieces then. I think that will improve the readability.

Reply:
Sounds good, I've moved this initialization out, Let me know if it's still too long.